File: csharp/src/detectors/server-side-request-forgery/ServerSideRequestForgery.cs
using System.Net.Http;

namespace SSRF
{
    class ServerSideRequestForgery
    {
        // {fact rule=server-side-request-forgery@v1.0 defects=1}
        public void ServerSideRequestForgeryNoncompliant(string hostname)
        {
            // Noncompliant: It is interacting with external network.
            Uri uri = new Uri(hostname);

            HttpClient client = new HttpClient();

            try
            {
                HttpResponseMessage res = client.GetAsync(uri).Result;
            }
            catch (Exception e)
            {
                System.Diagnostics.Debug.WriteLine(e);
            }
        }
        // {/fact}
        
        // {fact rule=server-side-request-forgery@v1.0 defects=0}
        public void ServerSideRequestForgeryCompliant(string hostname)
        {
            // Compliant: It is passing raw string.
            Uri uri = new Uri("string");

            HttpClient client = new HttpClient();

            try
            {
                HttpResponseMessage res = client.GetAsync(uri).Result;
            }
            catch (Exception e)
            {
                System.Diagnostics.Debug.WriteLine(e);
            }
        }
        // {/fact}
    }
}

File: ruby/src/detectors/server-side-request-forgery/server-side-request-forgery.rb
=begin
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
=end

# {fact rule=server-side-request-forgery@v1.0 defects=1}
require 'net/http'

def server_side_request_forgery_noncompliant
  url = params[:url]
  # Noncompliant: url is passing through `HTTP.get()`.
  Net::HTTP.get(url, "/index.html")
end
# {/fact}

# {fact rule=server-side-request-forgery@v1.0 defects=0}
def server_side_request_forgery_compliant
  url = params[:url]
  # Compliant: url is not passing through `HTTP.get()`.
  Net::HTTP.get("example.com", "/index.html")
end
# {/fact}
